lm_eval/models/hf_steered.py

```diff
@@ -0,0 +1,226 @@
+from contextlib import contextmanager
+from functools import partial
+from pathlib import Path
+from typing import Any, Callable, Generator, Union
+
+import torch
+from peft.peft_model import PeftModel
+from torch import Tensor, nn
+from transformers import PreTrainedModel
+
+from lm_eval.api.registry import register_model
+from lm_eval.models.huggingface import HFLM
+
+
+@contextmanager
+def steer(
+    model: Union[PreTrainedModel, PeftModel], hook_to_steer: dict[str, Callable]
+) -> Generator[None, Any, None]:
+    """
+    Context manager that temporarily hooks models and steers them.
+
+    Args:
+        model: The transformer model to hook
+        hook_to_steer: Dictionary mapping hookpoints to steering functions
+
+    Yields:
+        None
+    """
+
+    def create_hook(hookpoint: str):
+        def hook_fn(module: nn.Module, input: Any, output: Tensor):
+            # If output is a tuple (like in some transformer layers), take first element
+            if isinstance(output, tuple):
+                output = (hook_to_steer[hookpoint](output[0]), *output[1:])  # type: ignore
+            else:
+                output = hook_to_steer[hookpoint](output)
+
+            return output
+
+        return hook_fn
+
+    handles = []
+    hookpoints = list(hook_to_steer.keys())
+
+    for name, module in model.base_model.named_modules():
+        if name in hookpoints:
+            handle = module.register_forward_hook(create_hook(name))
+            handles.append(handle)
+
+    if len(handles) != len(hookpoints):
+        raise ValueError(f"Not all hookpoints could be resolved: {hookpoints}")
+
+    try:
+        yield None
+    finally:
+        for handle in handles:
+            handle.remove()
+
+
+@register_model("steered")
+class SteeredModel(HFLM):
+    def __init__(
+        self,
+        pretrained: str,
+        steer_path: str,
+        device: Union[str, None] = None,
+        **kwargs,
+    ):
+        """
+        HFLM with a steered forward pass.
+
+        To derive steering vectors from a sparse model loadable with sparsify or sae_lens,
+        provide the path to a CSV file with the following columns (example rows are provided below):
+
+        loader,action,sparse_model,hookpoint,feature_index,steering_coefficient,sae_id,description,
+        sparsify,add,EleutherAI/sae-pythia-70m-32k,layers.3,30,10.0,,,
+        sae_lens,add,gemma-scope-2b-pt-res-canonical,layers.20,12082,10.0,layer_20/width_16k/canonical,increase dogs,
+
+        To load steering vectors directly, provide the path to a pytorch (.pt) file with content in the following format:
+
+        [
+            {
+                "hookpoint": <str>,
+                "steering_vector": <torch.Tensor>,
+                "steering_coefficient": <float>,
```

Comment:
TODO there's likely no need for a coefficient here since we can save the vector with whatever multiple already applied. Gated on clamping decision.

Reply:
Keeping it to enable clamping to a fixed or maximum direction norm